### ISSUE TITLE
Issues on RDT Barcode Scan Function

### ISSUE THREAD

**Issue by ellyasafaa** [onaio/rdt-standard#747]
Please kindly find the image below for your reference. If you need to break down the issue and create new issue for these, please tag me on the issue.

![EoA_RDT Scan Function_1](https://user-images.githubusercontent.com/73918009/106087780-46095d80-6157-11eb-81ff-2c02d2988a2e.png)
![EoA_RDT Scan Function_2](https://user-images.githubusercontent.com/73918009/106087775-443f9a00-6157-11eb-8d64-8995edcac819.png)

**Comment by ellyasafaa** [comment]
@vincent-karuri @mwkhan1983

**Comment by vincent-karuri** [comment]
@ellyasafaa I think we discussed auto-detecting RDT type previously. For now it's a manual process that can be automated later on.

**Comment by vincent-karuri** [comment]
Manual capture was added as a timeout feature in case workers are struggling a lot with capturing the RDTs using automatic capture so as to allow them proceed with data collection.

I'm not sure that's something we want to turn off. We can make it a longer timeout however.

**Comment by ellyasafaa** [comment]
> @ellyasafaa I think we discussed auto-detecting RDT type previously. For now it's a manual process that can be automated later on.

Yes, I can recall clearly that. My apologies that at the time, I did not state clearly that the function is really critical for the apk. I wasn't aware of the purpose. In fact, the scanner is supposed to be more advance because it is supposed to be able to read the result also, right? How long will it take to make it automated?

> Manual capture was added as a timeout feature in case workers are struggling a lot with capturing the RDTs using automatic capture so as to allow them proceed with data collection.
> 
> I'm not sure that's something we want to turn off. We can make it a longer timeout however.

I get your point. However, we tried yesterday to the other RDT that is not included in part of the RDT Open Guideline, but the scanner is still able to read that, even it is wrongly pointed that as one of the RDT Open Guideline. I'm not referring only to that capture image function, but also to the scanner function, which is supposed to be able to detect the correct RDT and only RDT that is included in the RDT Open Guideline. The function of being able to take the picture of any part of RDT even any kind of picture that we should avoid. Everything that is not automatic should be avoided.

I have tried it as well in a dark-light condition room on 3 different devices, and it can read the RDT well. So, what causes the workers to struggle with capturing the RDTs?

**Comment by vincent-karuri** [comment]
Yes, the scanner interprets the reading behind the scenes. We still want to collect the user's interpretation for our own data analysis later on.

Not sure what is meant by `but the scanner is still able to read that`? Is this for manual capture? The purpose of manual capture is to allow capturing any RDT even when auto-capture fails. So yes, it will capture any RDT or image (even one which is not an RDT).

I would suggest that this is more a gap in training. Maybe you can observe how the workers perform capture vs what is recommended and adjust that?

**Comment by vincent-karuri** [comment]
@ellyasafaa ^

**Comment by ellyasafaa** [comment]
@vincent-karuri And in that case, any analysis will be done by SID personnel and all the data will be accessed only with authorization. And for now, there is no automatic feature of RDT interpretation, isn't it?

How long will it take to make the RDT type auto-detected by the app?

Yes, my concern is taking any picture, even if it's not RDT. I thought we could do something with the app to reduce the possibility of that. But, yeah, we still need the manual capture.

**Comment by vincent-karuri** [comment]
@pld are you aware of this `any analysis will be done by SID personnel and all the data will be accessed only with authorization`?

**Comment by vincent-karuri** [comment]
@ellyasafaa as I explained, auto-interpretation is done behind the scenes and not displayed to the user. This is to allow later comparison of user and machine interpretations.

**Comment by ellyasafaa** [comment]
@vincent-karuri So, how long will it take to make the RDT type auto-detected by the app?

**Comment by pld** [comment]
is RDT type stored in the barcode?

**Comment by vincent-karuri** [comment]
@pld That can be extracted from the device definition - I'm assuming rdt type means antigen/antibody.

Though I was asking about the flow where the chw has to input the result manually, even if we still perform auto-interpretation in the "back-end". We want to maintain this, right?

Do we still intend to perform analytics on the data?

**Comment by pld** [comment]
yes, auto interpretation won't happen on device, no one has calibrated this because there isn't a dataset I'm aware of

**Comment by vincent-karuri** [comment]
@ellyasafaa is it clear why we keep the RDT result question now? So while we can add a feature to detect the type: `Antigen/Antibody` we'll still need the user to fill in the result they read off the RDT.

**Comment by ellyasafaa** [comment]
@vincent-karuri Yes, of course. I am aware of the auto-interpret feature of the RDT results that have not been available in the RDT. Thank you for the explanation. The thing that I tried to emphasize before was indeed the type, not the results.

Also, I have been trying to use the RDT barcode scanner function in 3 types of RDT cassettes on Android 8 and 9. There were crashes that happened several times and it couldn't detect the RDT cassette automatically for the other 2 types although they have the same shapes. For Prevent, sometimes it can be detected automatically, but most of the time it couldn't. I thought the shape will determine how fast it will be to capture the image.

For additional info, the crash, says: Sorry, a fatal error occurred when loading the form. Please contact a developer for further assistance. Trying to upload the video here, but it doesn't work because the size is too big. I'm putting the videos in gdrive and will let you know after the upload is done.

**Comment by ellyasafaa** [comment]
@vincent-karuri Sent via email.

**Comment by vincent-karuri** [comment]
@ellyasafaa unless the RDT you're trying to capture is listed in the drop-down list, you will not be able to capture the RDTs automatically. The screen after scanning a barcode shows the RDT that the camera will be trying to capture once it's launched. A similar-looking RDT might be captured by chance but it's not assured to work.

In some of the cases (where we have crashes), I see the scanner is launching twice. Are you back-pressing after a successful scan by any chance?

**Comment by ellyasafaa** [comment]
@vincent-karuri Oh, I see. So, we have to include those RDTs in the list. Thanks for clearing these things up.

Nope, I didn't back-press the screen. I have tried it several times on both devices and it happened.

**Comment by ellyasafaa** [comment]
@vincent-karuri how if we want to add other RDTs to the drop-down list? What information needed?

**Comment by vincent-karuri** [comment]
@ellyasafaa but the scanner is launching twice, right? Which phones are affected?

**Comment by vincent-karuri** [comment]
> @vincent-karuri how if we want to add other RDTs to the drop-down list? What information needed?

For this, this [doc](https://docs.google.com/spreadsheets/d/1P9yCzXHoy9Dz0_hp43GX_qpOc6Ayvz_D2TCReNVN_n4/edit#gid=0) needs to be updated with the device configrations. And barcodes updated to reflect the changes for a specific RDT model.

**Comment by pld** [comment]
To be clear though these don’t get pulled in automatically, new configs need to be generated, and a new APL must be created to support new RDTs in the current state. We want this to be dynamic but that work is out of scope for now.

**Comment by ellyasafaa** [comment]
> @ellyasafaa but the scanner is launching twice, right? Which phones are affected?

Both, the Android 8 and Android 9. But on Android 8, there is no _Sorry, a fatal error occurred when loading the form. Please contact a developer for further assistance._ message.

**Comment by vincent-karuri** [comment]
@ellyasafaa What's the actual phone model, including Android version?

**Comment by ellyasafaa** [comment]
@vincent-karuri Redmi note 5 Pro Android 8.1.0 and Redmi Note 8 Android 9

**Comment by ellyasafaa** [comment]
@vincent-karuri issue on Xiaomi note 5 Pro Android 9. Can't go into the scanning page. Have installed OneScan in the device.

https://user-images.githubusercontent.com/73918009/106588060-efe45200-657c-11eb-8dd3-dc3e234d6e40.mp4

**Comment by vincent-karuri** [comment]
@ellyasafaa I'm not able to replicate this https://github.com/onaio/rdt-standard/issues/747#issuecomment-771540227 behaviour or the crashes. Is this the first time you're using the app on these devices? Are the issues affecting only the Xiaomi Redmi phones?

**Comment by ellyasafaa** [comment]
@vincent-karuri Since we only have Xiaomi Redmi phones here, I couldn't check it on another device brand, but on LG is fine. Yes, this is the first time we're using the app on the devices. I think it is important to solve these issues on this phone because Xiaomi and Redmi are the most common phone here.

**Comment by ellyasafaa** [comment]
@vincent-karuri In the barcode scanner, it will still be important that there is some checking in the app so if one RDT is selected and another is attempted to be scanned then this error is recognized and prevented.

**Comment by ellyasafaa** [comment]
@vincent-karuri please let us know more about how we could simply add additional RDTs from Indonesia and How can we also add the specific instructions for each RDT, including timings for incubation, etc?

**Comment by mwkhan1983** [comment]
@ellyasafaa 

> @vincent-karuri please let us know more about how we could simply add additional RDTs from Indonesia and How can we also add the specific instructions for each RDT, including timings for incubation, etc?

I have addressed addition of new RDTs in the last email sent. About specific instructions - if instructions are generic and not very specific to each RDT - then we can simply hardcode. If instructions are very specific to each RDT, then there is more work to be done to make it happen. Please share the instructions and we will evaluate the effort involved.

**Comment by ellyasafaa** [comment]
Sorry, for the sheets that you sent me before, I am only a commenter there, I thought you will add the column of the type and fill it based on this comment. You will extract it from the device definition?

> @pld That can be extracted from the device definition - I'm assuming rdt type means antigen/antibody.
> 
> Though I was asking about the flow where the chw has to input the result manually, even if we still perform auto-interpretation in the "back-end". We want to maintain this, right?
> 
> Do we still intend to perform analytics on the data?

**Comment by vincent-karuri** [comment]
@ellyasafaa @mwkhan1983 keep in mind that we were developing generic RDT job aids as part of this issue https://github.com/onaio/rdt-standard/issues/358.

**Comment by ellyasafaa** [comment]
@vincent-karuri Hi Vincent, could you enable the crashlytics on the new version of the APK? Found some crashes with Android 8 and 9.

**Comment by ellyasafaa** [comment]
@vincent-karuri @mwkhan1983 Also, Dr. Anu asked to put also the Prevent Malaria in the RDT Barcode scan function, could you assist us with that? The code can be pulled here: (https://github.com/cjpark87/rdt-scan/)

**Comment by ellyasafaa** [comment]
@vincent-karuri @mwkhan1983 Hi, I'm curious about the RDTScan Configuration. So, I thought at first, the size and shape will determine the capturing image function after scanning. But, it didn't work that way indeed.

So, there are some things that I'd like to clarify:
- After scanning the barcode, there will appear the information embedded on the barcode, including which type and brand of RDT that we are currently using to do the test. Will the results of the barcode scan determine the capturing image process? Are they connected? For example, if I scan the barcode and I got All Test RDT as the result, will the captured image process be automated only for the All Test RDT? Or, will the capturing image process be automated for all RDTs that have the RDTScan Configuration in the sheet you showed me?

- Secondly, assuming I have All Test Antigen RDT and in Antigen RDT, it only has two lines, right? The Control and Test Lines. It has exactly the same shape and the same result window as the Antibody one that we already have the configuration, the different thing is only the line. Could it be captured as automated as the Antibody one from the technical (or developer) perspective? And Could we use the same RDTScan Configuration for it?

Thank you

**Comment by ellyasafaa** [comment]
@vincent-karuri @mwkhan1983 Just wanna confirm so I didn't misunderstand. In the recent APK, has all the RDTScan Configuration in the sheet been merged there? Including the Prevent Malaria that CJ shared with us in the email? And the only one antigen RDT that has the configuration is JoysBio, right?

**Comment by mwkhan1983** [comment]
@ellyasafaa do we still have crash or clicking issues on Barcode scan with the latest release? If not, can we close this issue?

**Comment by ellyasafaa** [comment]
@vincent-karuri how is the autodetected RDT type going?
